pkg/activator/handler/context_handler.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2020 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"go.uber.org/zap"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/logging/logkey"
+	"knative.dev/serving/pkg/activator"
+	"knative.dev/serving/pkg/apis/serving/v1alpha1"
+	revisioninformer "knative.dev/serving/pkg/client/injection/informers/serving/v1alpha1/revision"
+	servinglisters "knative.dev/serving/pkg/client/listers/serving/v1alpha1"
+)
+
+type revisionKey struct{}
+type revIDKey struct{}
+
+// NewContextHandler creates a handler that extracts the necessary context from the request
+// and makes it available on the request's context.
+func NewContextHandler(ctx context.Context, next http.Handler) *ContextHandler {
+	handler := &ContextHandler{
+		nextHandler:    next,
+		revisionLister: revisioninformer.Get(ctx).Lister(),
+		logger:         logging.FromContext(ctx),
+	}
+
+	return handler
+}
+
+// ContextHandler enriches the request's context with structured data.
+type ContextHandler struct {
+	revisionLister servinglisters.RevisionLister
+	logger         *zap.SugaredLogger
+	nextHandler    http.Handler
+}
+
+func (h *ContextHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	namespace := r.Header.Get(activator.RevisionHeaderNamespace)
+	name := r.Header.Get(activator.RevisionHeaderName)
+	revID := types.NamespacedName{Namespace: namespace, Name: name}
+	logger := h.logger.With(zap.String(logkey.Key, revID.String()))
+
+	revision, err := h.revisionLister.Revisions(namespace).Get(name)
+	if err != nil {
+		logger.Errorw("Error while getting revision", zap.Error(err))
+		sendError(err, w)
+		return
+	}
+
+	ctx := r.Context()
+	ctx = logging.WithLogger(ctx, logger)
+	ctx = context.WithValue(ctx, revisionKey{}, revision)
```

Comment:
1. I am not sure about passing around revision (given that map lookup in lister might be faster than lookup on the linked list in context). 
2. The object might irrelevant by the time request returns, especially if it's a long running one?

Reply:
1. Yeah I don't think there's signifcant time to be gained here. It's worth noting that it's non-contentious in the context and that the revision will be sitting at a very early spot in the linkedlist, which should make the lookup fairly quick. We can potentially optimize further with a custom context here.
2. This is partially by design. I think if a request arrives in the system and has a valid revision, we should treat that revision object as immutable from thereon. I see it as a feature of this that the object cannot change in between handlers.

A meta point: I like that this enables us to do the error handling around revision fetching only once and not once per handler (which was arguably buggy before this change, some handlers didn't account for it or didn't properly test it at least).